File: Medium/684. Redundant Connection.py
from typing import List


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parents = list(range(len(edges) + 1))
        def find(child: int) -> int:
            if parents[child] != child:
                parents[child] = find(parents[child])
            return parents[child]
        def union(node1: int, node2: int) -> bool:
            parent1, parent2 = find(node1), find(node2)
            if parent1 == parent2:
                return False
            parents[parent1] = parent2
            return True
        for u, v in edges:
            if not union(u, v):
                return [u, v]

File: Medium/684. Redundant Connection.java
class Solution {
    class UnionFind {
        int[] parents;
        UnionFind(int n) {
            parents = new int[n + 1];
            for (int i = 1; i <= n; i++) {
                parents[i] = i;
            }
        }
        int find(int child) {
            if (parents[child] == child) {
                return child;
            }
            return parents[child] = find(parents[child]);
        }
        boolean union(int u, int v) {
            int uParent = find(u);
            int vParent = find(v);
            if (uParent == vParent) {
                return false;
            }
            parents[uParent] = vParent;
            return true;
        }
    }
    public int[] findRedundantConnection(int[][] edges) {
        final var UF = new UnionFind(edges.length);
        for (var edge: edges) {
            if (!UF.union(edge[0], edge[1])) {
                return edge;
            }
        }
        return null;
    }
}